lang/fpc/Portfile

```diff
@@ -134,34 +133,50 @@ if {${subport} eq "${name}"} {
     }
 
     switch ${build_arch} {
-        "x86_64" {
-            set b "ppcx64"
+        x86_64 {
+            set bootstrapCompiler ppcx64
+            set cpuTarget x86_64
+            set compiler ppcx64
+        }
+        arm64 {
+            set bootstrapCompiler ppcx64
+            set cpuTarget aarch64
+            set compiler ppca64
         }
         default {
-            set b "unsupported"
+            set bootstrapCompiler unsupported
+            set cpuTarget unsupported
+            set compiler unsupported
         }
     }
 
+    # os.major = 20 is macOS = 11 (BigSur)
+    if {${os.major} >= 20} {
+        set linkerOptions -XR${configure.sdkroot}
+    } else {
+        set linkerOptions ""
+    }
+
     worksrcdir          ${name}build-${version}/fpcsrc
-    build.env           PP=${workpath}/${b} \
+    build.env           PP=${workpath}/${bootstrapCompiler} \
                         PREFIX=${destroot}${fpcbasepath}
-    build.args          OPT="-ap -v0 -XR/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk"
-    build.target        all
+    build.args          OPT="-ap -v0 ${linkerOptions}"
+    build.target        all CPU_TARGET=${cpuTarget}
 
     destroot.env        {*}${build.env}
 
     # build the compiler utilities msgdif and msg2inc
     post-build {
         system -W ${worksrcpath}/compiler/utils \
-               "../ppcx64 -WM10.9 -Fu../../rtl/units/${build_arch}-darwin -ap -v0 \
-                  -XR/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk msgdif.pp && \
-                ../ppcx64 -WM10.9 -Fu../../rtl/units/${build_arch}-darwin -ap -v0 \
-                  -XR/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk msg2inc.pp"
+               "../${compiler} -WM${macosx_deployment_target} -Fu../../rtl/units/${cpuTarget}-darwin -ap -v0 \
```

Comment:
I like using the current system's deployment target as this will fix the builds on systems < 10.9, which I have had to manually fix before, as they were hard-coded to 10.9. I thought there was some reason you needed it to be 10.9, actually.

Reply:
Actually, I am not really sure about this. 10.7 may still break for two reasons. First, the bootstrap compiler is probably for 10.9 and up. There is one for 10.7, but it has version 3.0.0, which is not sufficient to build 3.2. This would require a additional step of building version 3.0.4 first. Second, some time ago, there is an issue with the assembler (https://bugs.freepascal.org/view.php?id=32324), which shows up with certain versions of XCode and has never been resolved. End of next week, I will have access and time to check these things on an actual 10.7 system. Although I wonder about whether it is worth the effort. The reason for adding this option was to silence some warnings, which were uncritical. What were your manual fixes?

Reply:
MacPorts has extremely strong support for older systems. As I recall, I only had to change 10.9 to 10.7 to make it work in the installed *.cfg (but I'd like to go back and look again).

Apple has two assemblers, the GNU "gas" and llvm's assembler driven by clang. The /usr/bin/as assembler driver defaulted to use "gas" until 10.9, and then Apple decided clang was robust enough, and changed the default, so the assembler driver would use "clang" preferentially.

MacPorts can install current clang versions on older systems, so MacPorts makes a slight tweak to the assembler driver installed by our cctools port (usually in /opt/local/bin/as) so that it will default to use some current version of clang on all systems, even back to 10.4 on Intel.

So that ticket you referenced in freepascal.org does not apply to MacPorts.

Great, eh?

Reply:
That gives me one more idea, at least as perspective. In case assembler and linker are available for i386 for all systems, that is all, what's needed for the i386 arch and with some extra effort even ppc. But let's get arm64 done first. Crosscompilers for windows are also an easy option as they use internal assembler and linker. Only Linux and related friends need external cross binutils.